lib/index.js

```diff
@@ -17,40 +28,585 @@ const PERMISSIONS = bedrock.config.permission.permissions;
+  }
+
+  logger.info('inserting account', {account});
+
+  // insert the account and get the updated record
+  const now = Date.now();
+  meta.created = now;
+  meta.updated = now;
+  meta.sequence = 0;
+  let record = {
+    id: database.hash(account.id),
+    meta,
+    account
+  };
+  const result = await database.collections.account.insert(
+    record, database.writeOptions);
+  record = result.ops[0];
+
+  // emit `postInsert` event with updated record data
+  eventData.account = bedrock.util.clone(record.account);
+  eventData.meta = bedrock.util.clone(record.meta);
+  await bedrock.events.emit('bedrock-identity.postInsert', eventData);
+
+  return record;
+});
+
+/**
+ * Retrieves an Account.
+ *
+ * @param actor the actor or capabilities for performing the action.
+ * @param id the ID of the Account to retrieve.
+ *
+ * @return a Promise that resolves to `{account, meta}`.
+ */
+api.get = brCallbackify(async ({actor, id}) => {
+  assert.string(id, 'account.id');
+
+  await _checkPermission({
+    actor,
+    permission: PERMISSIONS.ACCOUNT_ACCESS,
+    resource: [id]
+  });
+
+  const record = await database.collections.account.findOne(
+    {id: database.hash(id)}, {account: 1, meta: 1});
+  if(!record) {
+    throw new BedrockError(
+      'Account not found.',
+      'NotFound',
+      {id: id, httpStatusCode: 404, public: true});
+  }
+
+  _expandRoles(record.meta.sysResourceRole);
+
+  return record;
+});
+
+/**
+ * Retrieves the ID of the Account that manages the given identity or `null`
+ * if there is no managing account set.
+ *
+ * @param actor the actor or capabilities for performing the action.
+ * @param identity the ID of the identity to get the managing Account for.
+ *
+ * @return a Promise that resolves to the account ID or `null`.
+ */
+api.getManagerId = brCallbackify(async ({actor, identity}) => {
+  assert.string(identity, 'identity');
+
+  const record = await database.collections.identity.findOne({
+    id: database.hash(identity),
+    'meta.bedrock-account.account': {$exists: 1}
+  }, {'meta.bedrock-account.account': 1});
+  if(!record) {
+    return null;
+  }
+  return record.meta['bedrock-account'].account;
+});
+
+/**
+ * Retrieves all Accounts matching the given query.
+ *
+ * @param actor the actor or capabilities for performing the action.
+ * @param [query] the optional query to use (default: {}).
+ * @param [fields] optional fields to include or exclude (default: {}).
+ * @param [options] options (eg: 'sort', 'limit').
+ *
+ * @return a Promise that resolves to the records that matched the query.
+ */
+api.getAll = brCallbackify(async (
+  {actor, query = {}, fields = {}, options = {}}) => {
+  // TODO: move permission check to after query to allow users with
+  // more granular permissions to use this function
+  await _checkPermission({
+    actor,
+    permission: PERMISSIONS.ACCOUNT_ACCESS
+  });
+
+  const records = await database.collections.account.find(
```

Comment:
What's the advantage of using `forEach` here over `toArray`?

Reply:
Got no persuasive argument except that I think it's a good pattern which I think is useful for large collections to avoid creating massive array, then looping over massive array.